modules/http/http.c

```diff
@@ -491,6 +223,324 @@ http_dd_set_timeout(LogDriver *d, glong timeout)
+  curl_easy_setopt(self->curl, CURLOPT_SSL_VERIFYPEER, self->peer_verify ? 1L : 0L);
+
+  curl_easy_setopt(self->curl, CURLOPT_DEBUGFUNCTION, _curl_debug_function);
+  curl_easy_setopt(self->curl, CURLOPT_VERBOSE, (long) (debug_flag || trace_flag));
+
+  curl_easy_setopt(self->curl, CURLOPT_TIMEOUT, self->timeout);
+
+  if (self->method_type == METHOD_TYPE_PUT)
+    curl_easy_setopt(self->curl, CURLOPT_CUSTOMREQUEST, "PUT");
+}
+
+
+static struct curl_slist *
+_add_header(struct curl_slist *curl_headers, const gchar *header, const gchar *value)
+{
+  GString *buffer = scratch_buffers_alloc();
+
+  g_string_append(buffer, header);
+  g_string_append(buffer, ": ");
+  g_string_append(buffer, value);
+  return curl_slist_append(curl_headers, buffer->str);
+}
+
+static struct curl_slist *
+_format_request_headers(HTTPDestinationDriver *self, LogMessage *msg)
+{
+  struct curl_slist *headers = NULL;
+  GList *l;
+
+  headers = _add_header(headers, "Expect", "");
+  if (msg)
+    {
+      /* NOTE: I have my doubts that these headers make sense at all.  None of
+       * the HTTP collectors I know of, extract this information from the
+       * headers and it makes batching several messages into the same request a
+       * bit more complicated than it needs to be.  I didn't want to break
+       * backward compatibility when batching was introduced, however I think
+       * this should eventually be removed */
+
+      headers = _add_header(headers,
+                            "X-Syslog-Host",
+                            log_msg_get_value(msg, LM_V_HOST, NULL));
+      headers = _add_header(headers,
+                            "X-Syslog-Program",
+                            log_msg_get_value(msg, LM_V_PROGRAM, NULL));
+      headers = _add_header(headers,
+                            "X-Syslog-Facility",
+                            syslog_name_lookup_name_by_value(msg->pri & LOG_FACMASK, sl_facilities));
+      headers = _add_header(headers,
+                            "X-Syslog-Level",
+                            syslog_name_lookup_name_by_value(msg->pri & LOG_PRIMASK, sl_levels));
+    }
+
+  for (l = self->headers; l; l = l->next)
+    headers = curl_slist_append(headers, l->data);
+
+  return headers;
+}
+
+static void
+_add_message_to_batch(HTTPDestinationDriver *self, LogMessage *msg)
+{
+  if (self->body_template)
+    {
+      log_template_append_format(self->body_template, msg, &self->template_options, LTZ_SEND,
+                                 self->super.seq_num, NULL, self->request_body);
+    }
+  else
+    {
+      if (self->request_body->len)
+        g_string_append_c(self->request_body, '\n');
+      g_string_append(self->request_body, log_msg_get_value(msg, LM_V_MESSAGE, NULL));
+    }
+}
+
+static worker_insert_result_t
+_map_http_status_to_worker_status(glong http_code)
+{
+  worker_insert_result_t retval;
+
+  switch (http_code/100)
+    {
```

Comment:
The `0` is invalid http response code, I would not handle it as successful case (and now it possible to get `0` see my comment below.)